cppser/dog_helper.cpp

```diff
@@ -0,0 +1,154 @@
+            "考勤结束时间 > datetime('now', 'localtime') and "
+            "考勤结束时间 < datetime('now', 'localtime', '3 minutes')";
+        Statement stmt(conn, sql);
+        while (true) {
+            auto row = stmt.next();
+            if (!row)
+                break;
+            ans.push_back({
+                Clock::str2time(row->get<std::string>(0).substr(11)),
+                row->get<std::string>(1),
+                row->get<int>(2)
+            });
+        }
+        return ans;
+    }
+
+    void parse_url(const Configuration& config, std::string& host, std::string& url) {
+        // First, we will get the URL.
+        url = config["url_stu_new"];
+        // The host is matched with this raw regex
+        std::regex re(R"(\d+\.\d+\.\d+\.\d+)");
+        std::smatch matched;
+        if (!std::regex_search(url, matched, re))
+            throw std::logic_error("No host can be separated!");
+        host = matched[0];
+    }
+
+    nlohmann::json execute_request(
+        const Configuration& config,
+        const std::vector<Student>& absent,
+        const LessonInfo& lesson,
+        Logfile& logfile
+    ) {
+        namespace asio = boost::asio;
+        std::string url, host;
+        parse_url(config, host, url);
+        logfile << "host is " << host << "\nurl is " << url << std::endl;
+        // The request body
+        const std::string req_body = [&]{
+            nlohmann::json j;
+            j["ID"] = lesson.anpai;
+            j["date"] = GivenTimeClock()(lesson.endtime);
+            j["isloadfacedata"] = false;
+            j["faceversion"] = 2;
+            return j.dump();
+        }();
+        logfile << "req_body:\n" << req_body << std::endl;
+        // See example, sync HTTP client
+        asio::io_context ioc;
+        asio::ip::tcp::resolver resolver(ioc);
+        asio::ip::tcp::socket socket(ioc);
+        // RAII object to ensure log is flushed when error occurs.
+        LogSection log_section(logfile);
+        // Write the request.
+        asio::streambuf req;
+        std::ostream req_stream(&req);
+        req_stream << "POST " << url << " HTTP/1.1\r\n"
+            << "Content-Type: application/json\r\n"
+            << "Host: " << host << "\r\n"
+            << "Content-Length: " << req_body.size() << "\r\n"
+            << "Connection: close\r\n\r\n"
+            << req_body;
+        try {
+            asio::connect(socket, resolver.resolve(host, "http"));
+        } catch (const boost::system::system_error& ex) {
+            throw NetworkError(ex.what());
+        }
+        logfile << "Connected to the school server.\n";
+        try {
+            asio::write(socket, req);
+        } catch (const boost::system::system_error& ex) {
+            throw NetworkError(ex.what());
+        }
+        logfile << "Request written..\n";
+        // Start the receive section
+        asio::streambuf response;
+        boost::system::error_code ec;
+        asio::read(socket, response, ec);
```

Comment:
The receiving here takes about 2 seconds, which is kind of slow. Perhaps do a little research into buffering?